File: num_integration.exs
# Mariana Ramirez
# A01787819

defmodule NumericalIntegration do

  def f(x) do
    4.0 / (1.0 + x * x)
  end

  def sequential(n) do
    width = 1.0 / n

    sum =
      Enum.reduce(0..(n - 1), 0.0, fn i, acc ->
        mid = (i + 0.5) * width
        acc + f(mid)
      end)

    sum * width
  end

  def sum_range(start, finish, n) do
    width = 1.0 / n

    Enum.reduce(start..finish, 0.0, fn i, acc ->
      mid = (i + 0.5) * width
      acc + f(mid)
    end)
  end

  def parallel(n) do
    cores = System.schedulers_online()
    size = div(n, cores)

    sum =
      1..cores
      |> Enum.map(fn i ->
        start = (i - 1) * size

        finish =
          if i == cores do
            n - 1
          else
            i * size - 1
          end

        Task.async(fn -> sum_range(start, finish, n) end)
      end)
      |> Enum.map(fn task -> Task.await(task, :infinity) end)
      |> Enum.sum()

    sum * (1.0 / n)
  end

  def run(n) do
    seq = sequential(n)
    par = parallel(n)

    IO.puts("Sequential pi: #{seq}")
    IO.puts("Parallel pi: #{par}")
    IO.puts("Reference pi: 3.141592653589793")
  end
end

n =  case System.argv() do
    [value] -> String.to_integer(value)
    _ -> 1_000_000
  end

NumericalIntegration.run(n)

File: num_integration.cpp
/*
Mariana Ramirez
A01787819
 */

#include <iostream>
#include <vector>
#include <thread>

using namespace std;

double f(double x) {
    return 4.0 / (1.0 + x * x);
}

double computePiSequential(long n) {
    double width = 1.0 / n;
    double sum = 0.0;

    for (long i = 0; i < n; i++) {
        double mid = (i + 0.5) * width;
        sum += f(mid);
    }

    return sum * width;
}

void computeRange(long start, long end, long n, double &result) {
    double width = 1.0 / n;
    result = 0.0;

    for (long i = start; i <= end; i++) {
        double mid = (i + 0.5) * width;
        result += f(mid);
    }
}

double computePiParallel(long n, int numThreads) {
    vector<thread> threads;
    vector<double> results(numThreads);

    long rangeSize = n / numThreads;
    long start = 0;

    for (int i = 0; i < numThreads; i++) {
        long end;

        if (i == numThreads - 1) {
            end = n - 1;
        } else {
            end = start + rangeSize - 1;
        }

        threads.push_back(thread(computeRange, start, end, n, ref(results[i])));

        start = end + 1;
    }

    for (int i = 0; i < numThreads; i++) {
        threads[i].join();
    }

    double total = 0.0;

    for (int i = 0; i < numThreads; i++) {
        total += results[i];
    }

    return total * (1.0 / n);
}

int main() {
    long n = 100000000;
    int numThreads = thread::hardware_concurrency();

    double sequential = computePiSequential(n);
    double parallel = computePiParallel(n, numThreads);

    cout.precision(20);

    cout << "Sequential pi: " << sequential << endl;
    cout << "Parallel pi:   " << parallel << endl;
    cout << "Reference pi:  " << 3.14159265358979323846 << endl;

    return 0;
}